Audit3. Refactor council budget handling.

shamil-gadelshin opened the pull request:
We need to refactor the council budget handling to reduce code duplication.

#### Changes
- added new methods for the `CouncilBudgetManager` trait:
  - try_transfer
  - transfer
  - increase_budget
- changed the `bounty` palelt to use new methods

[Original issue](https://github.com/Joystream/joystream/issues/3496)

shamil-gadelshin committed:
runtime: wg: Add wg funding extrinsic with tests.

add `fund_working_group_budget`

shamil-gadelshin committed:
runtime: wg: Add benchmarking for extrinsic.

Add benchmarks and tests for the `fund_working_group_budget` extrinsic.

shamil-gadelshin committed:
runtime: Update weights for working group.

- Add weights for the `fund_working_group_budget` extrinsic.

shamil-gadelshin committed:
runtime: council: Add `fund_council_budget` extrinsic.

Add `fund_council_budget` extrinsic with tests, comments and benchmarking.

shamil-gadelshin committed:
Update chain-metadata.

shamil-gadelshin committed:
runtime: Update council and wg pallets weights.

shamil-gadelshin committed:
runtime: Add vesting pallet.

shamil-gadelshin committed:
Bump crate versions.

shamil-gadelshin committed:
runtime: Add vesting lock ID.

shamil-gadelshin committed:
Update chain metadata.

shamil-gadelshin committed:
Update chain metadata.

dobertRowneySr committed:
refactored validate_bag_params & bag change construction

dobertRowneySr committed:
fix clippy errors

dobertRowneySr committed:
fixed clippy errors

dobertRowneySr committed:
added empty creation params list check

dobertRowneySr committed:
added empty bag checks

dobertRowneySr committed:
cleaned content part

dobertRowneySr committed:
added bag creation with deletion prize now a parameter

dobertRowneySr committed:
added object creation

dobertRowneySr committed:
refactoring create_dyn_bag & upload_data_objects

dobertRowneySr committed:
added balance accounting

dobertRowneySr committed:
finished balance refactoring

dobertRowneySr committed:
clippy errors fixed

dobertRowneySr committed:
modified size check

dobertRowneySr committed:
remove optional deletio prize

dobertRowneySr committed:
fixed clippy warnings

dobertRowneySr committed:
fixed clippy warnings

dobertRowneySr committed:
fixing tests

dobertRowneySr committed:
all tests now passing

dobertRowneySr committed:
all content-tests now passing

dobertRowneySr committed:
refactoring

dobertRowneySr committed:
added bucket overflow check

dobertRowneySr committed:
objects to state

dobertRowneySr committed:
added net balance types

dobertRowneySr committed:
bag updated

dobertRowneySr committed:
added comments for future references

dobertRowneySr committed:
refactoring

dobertRowneySr committed:
refactor + added setters for bag

dobertRowneySr committed:
added bag operation

dobertRowneySr committed:
adding utility to construct objects

dobertRowneySr committed:
added utility function to construct objs to remove

dobertRowneySr committed:
fixed clippy errors

dobertRowneySr committed:
fixed clippy

dobertRowneySr committed:
added methods for list constructions

dobertRowneySr committed:
need to remove the net prize enum

dobertRowneySr committed:
added storage state mutation parts

dobertRowneySr committed:
finished storage bucket check

dobertRowneySr committed:
fixed todo on retrieve_bag

dobertRowneySr committed:
fixing storage bucket update algorithm

dobertRowneySr committed:
added distribution buckets to bag

dobertRowneySr committed:
fixed clippy errors

dobertRowneySr committed:
fixed clippy warnings

dobertRowneySr committed:
general cleanup

dobertRowneySr committed:
fixed tests

dobertRowneySr committed:
trying to debug tests

dobertRowneySr committed:
all tests now passing

dobertRowneySr committed:
fixed content side with tests passing

dobertRowneySr committed:
added event

dobertRowneySr committed:
added utility function in order to check balance

dobertRowneySr committed:
Btree set to reference

dobertRowneySr committed:
added docs

dobertRowneySr committed:
adding tests

dobertRowneySr committed:
fixed fixture delta

dobertRowneySr committed:
size used & objects used check added

dobertRowneySr committed:
added upload data objects test checks

dobertRowneySr committed:
added cases for error result in upload data obj fixture

dobertRowneySr committed:
added bag deletion prize to runtime

dobertRowneySr committed:
generated chain metadata

dobertRowneySr committed:
fixed benchmarks

dobertRowneySr committed:
added types

shamil-gadelshin committed:
runtime: storage: Convert bag deletion prize to storage value.

shamil-gadelshin committed:
runtime: Convert data object deletion prize to the storage value.

shamil-gadelshin committed:
runtime: Change buckets assignment on bag creation.

Changes:
- storage and content pallets
- removed random storage and distribution bucket assignment
- introduced storage and distribution buckets parameters on the dynamic bag creation
- introduced storage buckets checks

shamil-gadelshin committed:
runtime: storage: Add distribution buckets checks.

- add distribution buckets checks on dynamic bag creation.
- remove obsolete trait fields MaxRandomIterationNumber and Randomness

shamil-gadelshin committed:
runtime: content: Fix tests.

shamil-gadelshin committed:
runtime: storage: Add constraints for bag creation policy.

shamil-gadelshin committed:
runtime: content: Fix tests.

Commit:
Runtime: CuratorGroup permissions and Channel privilege_level

Commit:
Runtime: Video deletion by moderator/curator

Commit:
Runtime: Drop the outdated concept of "censoring" a video/channel

Commit:
Runtime: Channel deletion by curator

Commit:
Runtime: Introduce MaxKeysPerCuratorGroupPermissionsByLevelMap constraint

Commit:
Runtime: Replace "let mut channel = .." with ChannelById::mutate

Commit:
Runtime: Channel/Video visibility status

Commit:
Improve ModerationPermissionsByLevel type, add PausedChannelFunctionality enum

dobertRowneySr committed:
removed comment

dobertRowneySr committed:
refactor complex arg into one variable

dobertRowneySr committed:
renaming

Commit:
Runtime: Pause channel functionality

Commit:
https://github.com/Joystream/joystream/pull/3366 review comments part 1

shamil-gadelshin committed:
runtime: content: Add extrinsic.

- add `update_channel_transfer_status` extrinsic.

Commit:
https://github.com/Joystream/joystream/pull/3366 review comments part 2

Commit:
Channel/Video -> data objects association

shamil-gadelshin committed:
runtime: content: Add extrinsic.

- add `accept_channel_transfer` extrinsic with tests

shamil-gadelshin committed:
runtime: content: Bump version up to 4.1.0

Commit:
Runtime: New Channel/Video assets removal logic

dobertRowneySr committed:
clearer BagOPerationParams enum

dobertRowneySr committed:
fixed format in decl_storage

shamil-gadelshin committed:
runtime: content: Add ‘pending transfer’ barrier.

- modify extrinsics: add ‘no pending transfers’ checks for several extrinsics.

dobertRowneySr committed:
event emission upon dyn bag remval

dobertRowneySr committed:
remvode "magic" constants

Commit:
Runtime: Video/Channel assets deletion by curator

dobertRowneySr committed:
ensure type bag_id statically safe

shamil-gadelshin committed:
runtime: content: Add pending transfer barriers.

Add ‘pending transfer’ check barrier for extrinsics:
- claim_channel_reward
- edit_post_text
- delete_video
- update_moderator_set

dobertRowneySr committed:
inverted errors

shamil-gadelshin committed:
runtime: content: add channel transfer parameters validation.

dobertRowneySr committed:
added doc

dobertRowneySr committed:
typo

dobertRowneySr committed:
generated types & metadata

dobertRowneySr committed:
addressed last comment

dobertRowneySr committed:
commented storage benchmark

shamil-gadelshin committed:
runtime: content: Fix review comments.

shamil-gadelshin committed:
runtime: content: Remove member set validation.

Removed `validate_member_set` check for extrinsics:
- create_channel
- update_channel
- update_moderator_set

Initiated by review comment:
https://github.com/Joystream/joystream/pull/3432#discussion_r828832664

dobertRowneySr committed:
very useless comment edit in order to trigger vercel deployment

dobertRowneySr committed:
adedd DynBagCreationParameters + generated types

dobertRowneySr committed:
commented generate weights benchmark storage v2

dobertRowneySr committed:
fixed balance conversion

dobertRowneySr committed:
moved balance accounting before the mutation mark

dobertRowneySr committed:
removed Some after mutation mark

shamil-gadelshin committed:
Merge branch 'rhodes' into audit3-merge-rhodes

# Conflicts:
#	chain-metadata.json
#	runtime-modules/content/src/lib.rs
#	runtime-modules/content/src/tests/mock.rs
#	runtime-modules/storage/src/lib.rs
#	runtime/src/lib.rs
#	runtime/src/runtime_api.rs
#	scripts/generate-weights.sh
#	types/augment/all/types.ts

dobertRowneySr committed:
fixed dynamic bag id issue & moved mutation below mark

dobertRowneySr committed:
more explicative comment

mnaamani committed:
Merge pull request #3253 from ignazio-bovo/audit3_fix_CleanerStorage&ContentIntegration

Audit3 cleaner storage & content integration

shamil-gadelshin committed:
Merge branch 'audit3' into audit3-merge-rhodes

shamil-gadelshin committed:
Merge pull request #3445 from shamil-gadelshin/audit3-merge-rhodes

Audit3. Merge rhodes branch into audit3 branch.

Commit:
Merge branch 'audit3' into content-directory-new-permissions-model

Commit:
Add handling for paused VideoNftIssuance in create_channel and update_channel, fix tests

Commit:
Fix clippy warnings

Commit:
Update chain-metadata and types

Commit:
ChannelBagMissing - add debug_assert

Commit:
Video deletion: Fix mutation execution order

dobertRowneySr committed:
replaced Currency of with Balance of

shamil-gadelshin committed:
Merge branch 'audit3' into mainnet-vesting

# Conflicts:
#	Cargo.lock
#	chain-metadata.json
#	node/Cargo.toml
#	runtime/Cargo.toml
#	runtime/src/constants.rs
#	runtime/src/lib.rs
#	utils/chain-spec-builder/Cargo.toml

dobertRowneySr committed:
removed all reward account

dobertRowneySr committed:
all tests passing

shamil-gadelshin committed:
Merge branch 'audit3' into audit3-channel-transfer

# Conflicts:
#	runtime-modules/content/src/lib.rs

shamil-gadelshin committed:
runtime: content: Add test.

shamil-gadelshin committed:
runtime: content: Fix test.

dobertRowneySr committed:
generated types & metadata

shamil-gadelshin committed:
Merge branch 'audit3' into audit3-add-deletion-prize-storage-value

# Conflicts:
#	runtime-modules/content/src/lib.rs
#	runtime-modules/content/src/tests/nft/cancel_open_auction_bid.rs
#	runtime-modules/content/src/tests/nft/make_bid.rs
#	runtime-modules/storage/src/lib.rs
#	runtime-modules/storage/src/tests/fixtures.rs
#	runtime-modules/storage/src/tests/mod.rs

Commit:
Fix clippy warnings

dobertRowneySr committed:
deposit_to_creator -> deposit_to_channel_account

bedeho committed:
Merge pull request #3443 from ignazio-bovo/audit3_feature_ModuleOwnedAccounts

Audit3 feature module owned accounts

shamil-gadelshin committed:
Merge pull request #3145 from shamil-gadelshin/mainnet-vesting

Mainnet. Runtime: add vesting pallet.

shamil-gadelshin committed:
Merge branch 'audit3' into audit3-channel-transfer

# Conflicts:
#	runtime-modules/content/src/tests/channels.rs
#	runtime-modules/content/src/types.rs

shamil-gadelshin committed:
Merge branch 'audit3' into fund-council-wg

# Conflicts:
#	chain-metadata.json
#	runtime-modules/bounty/src/tests/mocks.rs
#	runtime-modules/council/src/benchmarking.rs
#	runtime-modules/council/src/lib.rs
#	runtime-modules/council/src/mock.rs
#	runtime-modules/council/src/tests.rs
#	runtime-modules/forum/src/mock.rs
#	runtime-modules/membership/src/tests/mock.rs
#	runtime-modules/proposals/codex/src/tests/mock.rs
#	runtime-modules/proposals/discussion/src/tests/mock.rs
#	runtime-modules/proposals/engine/src/tests/mock/mod.rs
#	runtime-modules/utility/src/tests/mocks.rs
#	runtime-modules/working-group/src/benchmarking.rs
#	runtime-modules/working-group/src/lib.rs
#	runtime-modules/working-group/src/tests/mock.rs
#	runtime/src/weights/council.rs
#	runtime/src/weights/working_group.rs

shamil-gadelshin committed:
Merge pull request #3432 from shamil-gadelshin/audit3-channel-transfer

Audit3. Runtime. Add channel transfer feature to the content pallet.

shamil-gadelshin committed:
Merge branch 'audit3' into audit3-add-deletion-prize-storage-value

# Conflicts:
#	runtime-modules/content/src/tests/channels.rs
#	runtime-modules/content/src/tests/merkle.rs
#	runtime-modules/content/src/tests/nft/accept_incoming_offer.rs
#	runtime-modules/content/src/tests/nft/buy_nft.rs

Commit:
Merge branch 'audit3' into content-directory-new-permissions-model

Commit:
Channel transfers: Update @joystream/types

Commit:
Merge branch 'content-directory-new-permissions-model' into channel-video-dataobjects-association

Commit:
Adjust existing types to match the runtime

Commit:
Merge branch 'content-directory-new-permissions-model' into channel-video-dataobjects-association

Commit:
Adjust existing types to match the runtime

Commit:
Address review comments

bedeho committed:
Merge pull request #3366 from Lezek123/content-directory-new-permissions-model

Content directory: New permissions model

shamil-gadelshin committed:
runtime: Add deletion prizes commitment to storage.

shamil-gadelshin committed:
Merge branch 'audit3' into audit3-add-deletion-prize-storage-value

# Conflicts:
#	runtime-modules/content/src/tests/channels.rs
#	runtime-modules/content/src/tests/fixtures.rs
#	runtime-modules/content/src/tests/videos.rs

shamil-gadelshin committed:
runtime: content: Remove dead code.

Commit:
Merge branch 'audit3' into channel-video-dataobjects-association

shamil-gadelshin committed:
Merge pull request #3317 from shamil-gadelshin/audit3-add-deletion-prize-storage-value

Audit3. Runtime: convert dynamic bag and data object deletion prizes to storage values.

shamil-gadelshin committed:
Revert "runtime: content: Remove member set validation."

This reverts commit bcc8751cd37d111c4dc09f6eae1a2d56a2f900a9.

# Conflicts:
#	runtime-modules/content/src/lib.rs
#	runtime-modules/content/src/tests/channels.rs

shamil-gadelshin committed:
runtime: content: Add `new_collaborators’ check.

- add `new_collaborators’ check for the `update_channel_transfer_status` extrinsic

Commit:
Address review comments

shamil-gadelshin committed:
Merge branch 'audit3' into audit3-channel-transfer-add-collab-checks

# Conflicts:
#	runtime-modules/content/src/tests/channels.rs

bedeho committed:
Merge pull request #3465 from shamil-gadelshin/audit3-channel-transfer-add-collab-checks

Audit3. Runtime. Restore `validate_member` check for the content pallet.

Commit:
Merge branch 'audit3' into channel-video-dataobjects-association

Commit:
Update types and chain-metadata

Commit:
Merge branch 'audit3' into channel-video-dataobjects-association

bedeho committed:
Merge pull request #3454 from Lezek123/channel-video-dataobjects-association

Channel/Video -> data objects association in the runtime

shamil-gadelshin committed:
Merge branch 'audit3' into audit3-user-defined-buckets

# Conflicts:
#	runtime-modules/content/src/lib.rs
#	runtime-modules/content/src/tests/channels.rs
#	runtime-modules/content/src/tests/fixtures.rs
#	runtime-modules/content/src/tests/videos.rs
#	runtime-modules/content/src/types.rs
#	runtime-modules/storage/src/lib.rs
#	runtime-modules/storage/src/random_buckets/storage_bucket_picker.rs
#	runtime-modules/storage/src/tests/fixtures.rs
#	runtime-modules/storage/src/tests/mod.rs

shamil-gadelshin committed:
runtime: storage: Fix review comments.

- refactor storage and distribution buckets validation on  the dynamic bag creation.

shamil-gadelshin committed:
Merge branch 'audit3' into audit3-user-defined-buckets

# Conflicts:
#	runtime-modules/content/src/tests/channels.rs
#	runtime-modules/content/src/tests/fixtures.rs
#	runtime-modules/content/src/tests/videos.rs

shamil-gadelshin committed:
Merge pull request #3348 from shamil-gadelshin/audit3-user-defined-buckets

Audit3. Convert random buckets selection to user defined buckets.

shamil-gadelshin committed:
Merge pull request #3068 from shamil-gadelshin/fund-council-wg

runtime: Add funding for working groups and concil by members.

shamil-gadelshin committed:
runtime: Enrich CouncilManager trait.

shamil-gadelshin committed:
runtime: Add tests for CouncilManager.

bedeho commented on runtime-modules/common/src/council.rs:
```diff
@@ -1,12 +1,29 @@
+    fn try_transfer(account_id: &AccountId, amount: Balance) -> DispatchResult;
+
+    /// Remove some balance from the council budget and transfer it to the account. Infallible.
+    fn transfer(account_id: &AccountId, amount: Balance) {
```
`withdraw` is a bit more explicit

bedeho replied:
Can you also explicitly say that if the budget does not cover full withdrawal, there is no side-effect.

shamil-gadelshin replied:
Addressed both comments: https://github.com/Joystream/joystream/pull/3516/commits/ad221feb144e467f4422acd273b058fbc6692ff1

bedeho reviewed (commented):
Nice work, just a cosmetic tweak.

shamil-gadelshin committed:
runtime: council: Refactor CouncilBudgetManager.

- rename try_transfer and transfer to try_withdraw and withdraw
- add more comments

bedeho reviewed (approved):
LGTM.

shamil-gadelshin commented:
Superceded by https://github.com/Joystream/joystream/pull/3518
